### DISCUSSION TITLE
How can I prioritize breaking out of a "repeat", when continued repeating is also valid?

### DISCUSSION THREAD

**Discussion by Hubro** [tree-sitter/tree-sitter#2410] [Q&A] [answered]
I'm trying to improve the Robot tree-sitter grammar by adding support for parsing the built-in conditional keywords.

Basically I'm trying to parse this:

```robot
Example
    IF    ${rc} > 0
        Some keyword
        Another keyword
    END
```

This is my attempt at matching the conditional statement:

```javascript
    conditional_statement: $ => seq(
      "IF",
      $._separator,
      $.argument,
      $._line_break,
      repeat1(choice($.statement, $._empty_line)),
      "END",
    ),
```

(For context, [here is the full grammar](https://github.com/Hubro/tree-sitter-robot/blob/master/grammar.js))

The issue is, `END` is technically a completely valid statement in Robot Framework. Any statement in Robot is a keyword followed by 0 or more arguments. A keyword can be literally any string, *except* for the built-in constructs like `IF`, `END`, `FOR` etc.

This means that with the above code, the `END` is parsed as just another normal statement by the `repeat1`, and the parse output eventually ends with `(MISSING "END")`

The ["exclude" rule](https://github.com/tree-sitter/tree-sitter/pull/246) would have been perfect in this situation, since I could just exclude built-in keywords like "IF" and "END" from being valid keyword names, but that PR was abandoned :disappointed: 

I have peppered my code with every combination of `conflicts`, `prec.left` and `prec.dynamic` I can imagine, but nothing has had any impact.

Is there any way I can get Tree-sitter to prioritize matching something *after* the `repeat` rule rather than keep repeating?

**Comment by amaanq** [comment]
tree-sitter will prioritize matching literals over a regex that could match END, however, if "END" is a valid choice in statement then that can be matched too

**Comment by Hubro** [comment, marked as answer]
Dang, turns out I just banged my head against this problem for too long. After a good night's sleep, it seemed obvious immediately. This grammar matches leading indentation as a token, and I forgot to match one before the `"END"` rule. Replacing `"END"` with `seq($._indentation, "END")` fixes the problem.
